codex/purchasing/states/submitted.nim

```diff
@@ -30,12 +30,12 @@ method run*(
     requestId = purchase.requestId
 
   proc wait() {.async.} =
-    let done = newFuture[void]()
+    let done = newAsyncEvent()
     proc callback(_: RequestId) =
-      done.complete()
+      done.fire()
```

Comment:
Is the callback supposed to be called more than once, if not this is probably an issue somewhere else?

Reply:
Agreed. This smells like a subscription callback or ethers error, or maybe the event is indeed being fired more than once in the contract.

The root issue will be masked by this change. I don't think it's a big deal personally, but it could potentially hide other issues.

Reply:
Agreed that it should not be called more than once. However it might be called more than once if the JSON-RPC provider decides to invoke the callback more than once for whatever reason (chain reorg?), so we'd better not crash if it does :smile: